clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -267,50 +173,77 @@ public T get() throws InterruptedException, ExecutionException {
     @Override
     public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException,
             TimeoutException {
-        SingleWaiter<T> waiter = new SingleWaiter<>();
-        addWaiter(waiter);
-        return waiter.await(timeout, unit);
+        try {
+            return completableFuture.get(timeout, unit);
+        } catch (ExecutionException e) {
+            maybeRewrapAndThrow(e.getCause());
+            throw e;
+        }
     }
 
     /**
      * Returns the result value (or throws any encountered exception) if completed, else returns
      * the given valueIfAbsent.
      */
     @Override
-    public synchronized T getNow(T valueIfAbsent) throws InterruptedException, ExecutionException {
-        if (exception != null)
-            wrapAndThrow(exception);
-        if (done)
-            return value;
-        return valueIfAbsent;
+    public synchronized T getNow(T valueIfAbsent) throws ExecutionException {
+        try {
+            return completableFuture.getNow(valueIfAbsent);
+        } catch (CompletionException e) {
+            maybeRewrapAndThrow(e.getCause());
+            // Note, unlike CF#get() which throws ExecutionException, CF#getNow() throws CompletionException
+            // thus needs rewrapping to conform to KafkaFuture API, where KF#getNow() throws ExecutionException.
+            throw new ExecutionException(e.getCause());
+        }
     }
 
     /**
      * Returns true if this CompletableFuture was cancelled before it completed normally.
      */
     @Override
     public synchronized boolean isCancelled() {
-        return exception instanceof CancellationException;
+        if (isDependant) {
+            Throwable exception;
+            try {
+                completableFuture.getNow(null);
+                return false;
+            } catch (Exception e) {
+                exception = e;
+            }
+            return exception instanceof CompletionException
+                    && exception.getCause() instanceof CancellationException;
```

Comment:
I am curious here. Would it be possible to let the `CancellationException` propagate from the upstream future to this one when the upstream future is cancelled?

Reply:
The historical API of `KafkaFuture` on cancellation of an upstream future (e.g. `thenApply()`) was for both futures to return true from `isDone()`, `isCancelled()` and `isCompletedExceptionally()`, and for both `getNow` and `get()` to throw a `CancellationException`. 

If the new implementation of `isCancelled()` just does `return completableFuture.isCancelled();` then for a dependent future `isCancelled()` would return false because it's exception is a `CompletionException` wrapping a `CancellationException` and `CompleteableFuture#isCancelled()` just checks for the exception being `CancellationException`. So we need to do something different for cancellation of dependent futures. 

Because `KafkaFutureImpl.thenApply()` doesn't itself keep track of waiters any more it's not possible to do anything clever in the implementation of `cancel()`, so we have to do something in `isCancelled()`.

So to (hopefully) answer your question: It does propagate already, but the APIs differ and what we're doing here is to maintain the existing semantics of `KafkaFuture`.

Reply:
Oh.. I see. Thanks for the clarification. It would be great if you could add a comment which explains this because this is not obvious.

Reply:
Added. Let me know whether it makes sense to you. Also: this is covered by the test which passes with the old and new implementations.